### PR TITLE
Migrate query api to rust

### PR DESCRIPTION
**Pull request by ruokun-niu**
# Description

Migrate Query API to Rust

## Type of change

- This pull request adds or changes features of Drasi and has an approved issue (issue link required).

Fixes: #issue_number

### CODE REVIEW

File: infrastructure/comms-abstractions/src/comms.rs [hunk 1]
```diff
@@ -36,5 +36,5 @@ pub trait Publisher {
         &self,
         data: Value,
         headers: Headers,
-    ) -> Result<(), Box<dyn std::error::Error>>;
+    ) -> Result<reqwest::Response, Box<dyn std::error::Error>>;
```
Comment (danielgerlag) [inline infrastructure/comms-abstractions/src/comms.rs:39]: I don't think we should include `reqwest::Response` as part of the API of drasi comms.  It tightly couples us to a specific library.

File: sources/shared/query-api/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+    let mut headers_map = std::collections::HashMap::new();
+    _ = match headers.get("traceparent") {
+        Some(tp) => match tp.to_str() {
+            Ok(tp) => {
+                headers_map.insert("traceparent".to_string(), tp.to_string());
+            }
+            Err(e) => {} // 
+        },
+        None => {}
```
Comment (danielgerlag) [inline sources/shared/query-api/src/main.rs:113]: We could log a warning if the trace headers are missing.

**Review by danielgerlag — CHANGES_REQUESTED** [review]

File: infrastructure/comms-abstractions/src/comms.rs (hunk 1 above)
Comment (ruokun-niu) [reply infrastructure/comms-abstractions/src/comms.rs:39]: How about a custom struct that contains the status code and/or the response body? Something like:
```rust
pub struct HttpResponse {
    pub status_code: u16,
    pub body: String,
}
```

Comment (danielgerlag) [reply infrastructure/comms-abstractions/src/comms.rs:39]: That might make sense for the `Invoker`, but the publisher implementation could be one that does not use Http, if we choose to not use Dapr in the future, many messages brokers have their own protocols.
I think if a non success code comes from the Dapr sidecar, the publish was unsuccessful, and we just return an error, and if the body contains details, we include that in the error result.

Comment (ruokun-niu) [reply infrastructure/comms-abstractions/src/comms.rs:39]: that makes sense. for a successful service invocation, maybe we can just return the body as Bytes?

Comment (ruokun-niu) [reply infrastructure/comms-abstractions/src/comms.rs:39]: oh actually this is the publisher, we don't really need to return anything if the publish is successful

Comment (ruokun-niu) [reply infrastructure/comms-abstractions/src/comms.rs:39]: Sorry I was think about the wrong thing

File: infrastructure/comms-abstractions/src/comms.rs [hunk 3]
```diff
@@ -26,7 +26,7 @@ pub trait Invoker {
         app_id: String,
         method: String,
         headers: Headers,
-    ) -> Result<(), Box<dyn std::error::Error>>;
+    ) -> Result<serde_json::Value, Box<dyn std::error::Error>>;
```
Comment (danielgerlag) [inline infrastructure/comms-abstractions/src/comms.rs:29]: We can't assume the response will always be json

**Review by danielgerlag — APPROVED** [review]

**Pull request merged**